src/transaction.cc

```diff
@@ -548,20 +548,54 @@ int Transaction::addRequestHeader(const std::string& key,
 
     if (keyl == "cookie") {
         size_t localOffset = m_variableOffset;
+        size_t pos;
         std::vector<std::string> cookies = utils::string::ssplit(value, ';');
         for (const std::string &c : cookies) {
-            std::vector<std::string> s = utils::string::split(c,
-               '=');
-            if (s.size() > 1) {
-                if (s[0].at(0) == ' ') {
-                    s[0].erase(0, 1);
+            // skip empty substring, eg "Cookie: ;;foo=bar"
+            if (c.length() == 0) {
+                localOffset++; // add length of ';'
+                continue;
+            }
+
+            // find the first '='
+            pos = c.find_first_of("=", 0);
+            std::string ckey = "";
+            std::string cval = "";
+
+            // if the cookie doesn't contains '=', its just a key
+            if (pos == std::string::npos) {
+                ckey = c;
+            }
+            // else split to two substrings by first =
+            else {
+                ckey = c.substr(0, pos);
+                // value will contains the next '=' chars if exists
+                // eg. foo=bar=baz -> key: foo, value: bar=baz
+                cval = c.substr(pos+1, c.length());
+            }
+
+            // handle cookie only if the key is not empty
+            if (ckey.length() > 0) {
+                // ltrim the key - following the modsec v2 way
+                while (ckey.length() > 0 && ckey.at(0) == ' ') {
+                    ckey.erase(0, 1);
+                    localOffset++;
+                }
+
+                // if the key is empty (eg: "Cookie:   =bar;" skip it
+                if (ckey.length() == 0) {
+                    localOffset = localOffset + c.length() + 1;
+                    continue;
```

Comment:
The flow here is more complicated than it needs to be.  Consider:
1) move the left-trimming while loop (currently lines 580-583) to be right after line 576
2) follow that with a single 'if (ckey.empty()) {' where the 'if' block has what's currently at line 587 and the 'else' block has the 591-598 content.

Reply:
Changed in e14f3dc.